### DISCUSSION TITLE
posts/mimic-go-defer-in-csharp/

### DISCUSSION THREAD

**Discussion by giscus** [BYJRK/byjrk.github.io#13] [Announcements]
# posts/mimic-go-defer-in-csharp/

Go 语言中的 defer 关键字非常好用，可以用来释放资源，关闭文件等。本文介绍了如何在 C# 中模拟 Go 的 defer 关键字，并将其用于 WPF 客户端开发。

https://blog.coldwind.top/posts/mimic-go-defer-in-csharp/

**Comment by baciZzz** [comment]
很好用的语法，特别是在使用Com对象的时候不需要一直写try-catch-finally，来保证com对象的释放了
```c#
//Control是扩展方法，返回一个IDisposable对象，Dispose会释放Com对象 => Marshal.ReleaseComObject(Value)
using var _ = new MapDocumentClass().Control(out IMapDocument mapDocument1);
using var _ = new MapDocumentClass().Control(out IMapDocument mapDocument2);

//dosomething
```

**Comment by qiangboy** [comment]
实现了IDisposable的这个类，会自动垃圾回收吗？

**Reply by BYJRK** [reply]
`IDisposable` 与 GC 垃圾回收其实并没有什么必然的关系。一般来说，如果一个类中包含非托管资源，那么如果不正确处理，可能会影响到这个类被 GC 回收。因此，通常做法是这个类实现上述接口，并在 `Dispose` 方法中处理非托管资源的释放。

而对于文件的操作，`Dispose` 通常也会保证文件的内容被正确保存，文件指针也会被释放掉。

但这并不意味着我们一定要在 `Dispose` 方法中添加只与垃圾回收相关的逻辑，就好比我们可以在析构函数中添加 `Console.WriteLine` 一样，我们想怎么做就怎么做，编译器管不着。比如本文的例子，我们就在 `Dispose` 中进行回调函数的调用，从而进行善后。

**Comment by emako** [comment]
我顺便把实现publish到nuget了：https://www.nuget.org/packages/Defer/。另外加了个readonly ref struct的实现

**Reply by emako** [reply]
https://www.nuget.org/packages/Defer/

**Reply by BYJRK** [reply]
优秀👍我前些天无意中在github feed里面居然刷到了
